Zend/zend_builtin_functions.c

```diff
@@ -2444,6 +2462,23 @@ ZEND_API void zend_fetch_debug_backtrace(zval *return_value, int skip_last, int
 		frameno++;
 		array_init(&stack_frame);
 
+		if (!ptr->func && ptr->prev_execute_data && Z_OBJ(ptr->This)) {
+			if (Z_OBJCE(ptr->This) == zend_ce_generator) {
+				zend_generator *generator = (zend_generator *) Z_OBJ(ptr->This);
+				zend_generator *root = (generator->node.children < 1 ? generator : generator->node.ptr.leaf)->node.ptr.root;
+				zend_execute_data *prev = ptr->prev_execute_data;
+				if (generator->node.parent != root) {
+					do {
+						generator->execute_data->prev_execute_data = prev;
+						prev = generator->execute_data;
+						generator = generator->node.parent;
+					} while (generator->node.parent->node.parent != root);
+				}
+				generator->execute_data->prev_execute_data = prev;
+				ptr = generator->execute_data;
+			}
+		}
```

Comment:
This code looks fully copy/paste, maybe stick it in a function (here or as a zend_generator_ API)?